diagnostics/lib.rs

```diff
@@ -0,0 +1,292 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+#![deny(unused_must_use)]
+#![deny(elided_lifetimes_in_paths)]
+
+use std::{
+    collections::{HashMap, HashSet},
+    hash::{DefaultHasher, Hash, Hasher},
+    path::PathBuf,
+    sync::{Mutex, MutexGuard},
+};
+
+use error::TypeDBError;
+use serde_json::Value as JSONValue;
+
+use crate::metrics::{
+    ActionKind, ActionMetrics, DatabaseMetrics, ErrorMetrics, LoadKind, LoadMetrics, ServerMetrics, ServerProperties,
+};
+
+pub mod diagnostics_manager;
+pub mod metrics;
+mod monitoring_server;
+mod reporter;
+mod version;
+
+type DatabaseHash = u64;
+type DatabaseHashOpt = Option<u64>;
+
+#[derive(Debug)]
+pub struct Diagnostics {
```

Comment:
`Diagnostics` consists of a number of `Properties` and `Metrics`. The model of these metrics is a replication of what we had in 2.x, but with slightly different names. 

The synchronization is placed here at the moment. 
We are still not fully interested in preserving the balance between `load_metrics` and `action_metrics`, so it's fine if they are updated and cleared on snapshotting in parallel, these losses are minimal and not critical. 

However, for simplicity, I placed the mutexes on the whole metrics now, which means that two transactions against two different databases will share a single mutex when they report their actions and errors. It's not ideal, but I decided to start with this approach, leaving the `***Metrics` structs unaware of the cruel concurrent world. This might be a performance issue!